### DISCUSSION TITLE
tutorials/export/exporting_for_web

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#24] [User-contributed notes]
# tutorials/export/exporting_for_web

HTML5 export allows publishing games made in Godot Engine to the browser. This requires support for WebAssembly, WebGL and SharedArrayBuffer in the user's browser. WebGL version: Godot 4.0 and late...

https://docs.godotengine.org/en/latest/tutorials/export/exporting_for_web.html

**Comment by Swarkin** [comment]
Since Godot 4.3, projects can now be exported with a single thread.
See https://godotengine.org/article/progress-report-web-export-in-4-3/

**Comment by lakido-sudo** [comment]
I have a problem with "...ArrayBufferView larger than 2 GB" =(

**Reply by Calinou** [reply]
This should be fixed in 4.4.dev2 and later: https://github.com/godotengine/godot/issues/96068

**Comment by ProbablyIdiot** [comment]
I get hundreds of script failed to load errors, when running from itch.io or localhost. I can access the leel select screen, but the physics doesn't work when I try to load a level (or it's just frozen). https://drive.google.com/file/d/1BO99uabvns79U2KJdD7ggo8ZD1bWaHm0/view?usp=sharing

**Reply by ProbablyIdiot** [reply]
(I forgot to add, theres many more errors than this, I just couldn't fit them on.

**Reply by Calinou** [reply]
Please ask support questions on other community platforms: https://godotengine.org/community/

**Reply by Swarkin** [reply]
It looks like you have an addon that doesn't work on web.

**Comment by shahriarlabib000** [comment]
So I'm on 4.4dev3 on Android editor and I wanted my game to play in portrait mode in browser. Couldn't figure out how. I set it to portrait in the project settings too

**Reply by shahriarlabib000** [reply]
Ahh it was the sites fault , sorry

**Comment by osmanyasar** [comment]
You can run your projects on localhost or https protocol using this tool.
https://github.com/osmanyasar/Godot-WebGL-Runner

**Reply by Swarkin** [reply]
Worth noting: You can host a simple web server by going into your export directory and typing `python -m http.server` in the command line if you have python installed already. Then visit the link it shows you and click on the .html to test out the game

**Reply by osmanyasar** [reply]
Yes, I know that. I'm approaching it from my point of view. I'd like to point out that I'm a 3D artist. My software knowledge is at a basic level. I work with a pc and a laptop on my desk. I got an HTTPS error while trying to open my build on the laptop and I solved it with another python script. I wanted to simplify it thinking that others who know basic software like me need it. I can understand that not everyone needs it.

**Reply by felipebottega** [reply]
@Swarkin great tip, thx!

**Comment by Kral-bohu** [comment]
**When you embed the game to your website using iframe, it'll jump ("scroll" down) to the iframe position when game loads.**
Besause you may not have your iframe on top of page or it loads for long time, you perhaps don't want this. **How to fix:** Go to generated html. Search for *focusCanvas* and you should find line starting with *const GODOT_CONFIG*. Change value of *focusCanvas* to *false* (*"focusCanvas":false*). **Fixed!**

**Comment by GeorgeFought** [comment]
Using 4.4.1 and macos the exporter now won't create the zip file. Macos will always create a directory inside of the zip so sites like itch.io won't be able to see the index.html. Any way to solve this?

**Reply by Calinou** [reply]
> Macos will always create a directory inside of the zip so sites like itch.io won't be able to see the index.html.

What do you mean? Are you referring to the `_MACOSX` folder? You can remove it, it's not needed for anything outside of macOS (it only stores things like Finder metadata).

**Reply by GeorgeFought** [reply]
no I mean like this:

directory:
    All export files (Even if I zip here)

the zip will have this same structure when you unzip. That's the same result if I zipped the directory. So no matter what I can't get the zip to work correctly for html exports for itch.io. I never ran into this issue with 4.3, because it exported as a zip.

**Reply by WolfgangSenff** [reply]
Can't you just export it the normal way and then zip it manually yourself, rather than exporting as a zip? That's generally how I've handled it in the past, and it works fine. It's easy enough to write a script which will zip it and remove the __MACOSX and .ds_store folders automatically as well, or for that matter move files around if for some reason you need to.

**Reply by GeorgeFought** [reply]
I export the normal way and try to zip the files.

The zip file has the files i exported (.html etc) inside of a new folder (I don’t create). So when you open the folder it looks like this:

New.zip
-some new directory/folder I don’t create named new (whatever the .zip is named)
— my files

so itch.io can’t see the index.html because it’s looking at the root level, not some unknown new file that shouldn’t be there.

the current work around is to export on my windows pc and zip there. When I zip the files on windows, windows doesn’t create a new folder inside the zip which contains the files. The files in the windows zip are at the root level of the zip.

I’m not sure why Mac is creating a new folder inside the zip to house the files. Everything I’ve read online shows it shouldn’t be, but it does no matter how I try to zip the files.

**Reply by jamesoncomputers** [reply]
Sounds like zipping software issue. Check the software for options to get rid of that directory, ask AI about it specifically, bet it will help you.

**Comment by noidexe** [comment]
> The project will be paused by the browser when the tab is no longer the active tab in the user's browser. This means functions such as _process() and _physics_process() will no longer run until the tab is made active again by the user (by switching back to the tab). This can cause networked games to disconnect if the user switches tabs for a long duration.

> This limitation does not apply to unfocused browser windows. Therefore, on the user's side, this can be worked around by running the project in a separate window instead of a separate tab.

This can happen even with separate windows if the browser window is fully obscured by other windows. Ensure that at least a tiny portion of the browser window is active. 

Tested with Firefox 138.0.3 (64-bit) and Chrome 136.0.7103.93 (64-bit) on Windows 11 24H2

**Comment by viktor-ferenczi** [comment]
Important gotcha on Firefox: 

Do **not** invoke `DisplayServer.clipboard_get()` frequently.

It would cause Firefox to show a popup menu with a single `Paste` item all the time, making the application unusable.

Chrome/Chromium just raises the security question of whether to allow clipboard usage, then it works fine.

Best to avoid this call entirely on the Web platform. Call it only once if the user presses Ctrl-V or a Paste button. The only drawback is you cannot properly disable the Paste button if the clipboard content is not suitable, but that's not a big deal.

**Comment by SriramAnanth2912** [comment]
Why am I getting unable to fetch error (CORS error) when opening the .html file in my browser.
It is giving CorsDisabledScheme error.
Actually I made mine as PWA with "ensure cross-origin isolation" still having this error.
Do I have to enable thread support?
<img width="1915" height="860" alt="Screenshot 2025-10-22 215750" src="https://github.com/user-attachments/assets/73e2b3a8-94bb-4f18-87e8-3dc9c3f467d0" />

**Reply by Calinou** [reply]
You need to host a local web server, as running a web export through a `file://` URL is not supported.

You can use the Python SimpleHTTPServer as described in this page, but I've also had success with https://github.com/TheWaWaR/simple-http-server personally.

**Reply by SriramAnanth2912** [reply]
Thank you I will look into that.
Meanwhile I successfully launched my first game in itch.io
